pkg/promtail/targets/gcplog/target.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/relabel"
+)
+
+var (
+	gcplogEntries = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "gcplog_target_entries_total",
+		Help:      "Help number of successful entries sent to the gcplog target",
+	}, []string{"project"})
+
+	gcplogErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "promtail",
+		Name:      "gcplog_target_parsing_errors_total",
+		Help:      "Total number of parsing errors while receiving gcplog messages",
+	}, []string{"project"})
+)
+
+// GcplogTarget represents the target specific to GCP project.
+// It collects logs from GCP and push it to Loki.
+type GcplogTarget struct {
+	logger        log.Logger
+	handler       api.EntryHandler
+	config        *scrapeconfig.GcplogTargetConfig
+	relabelConfig []*relabel.Config
+	jobName       string
+
+	// lifecycle management
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// pubsub
+	ps   *pubsub.Client
+	msgs chan *pubsub.Message
+}
+
+// NewGcplogTarget returns the new instannce of GcplogTarget for
+// the given `project-id`. It scraps logs from the GCP project
+// and push it Loki via given `api.EntryHandler.`
+// It starts the `run` loop to consume log entries that can be
+// stopped via `target.Stop()`
+func NewGcplogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GcplogTargetConfig,
+) (*GcplogTarget, error) {
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	ps, err := pubsub.NewClient(ctx, config.ProjectID)
+	if err != nil {
+		return nil, err
+	}
+
+	target, err := newGcplogTarget(logger, handler, relabel, jobName, config, ps, ctx, cancel)
+	if err != nil {
+		return nil, err
+	}
+
+	go target.run()
+
+	return target, nil
+}
+
+func newGcplogTarget(
+	logger log.Logger,
+	handler api.EntryHandler,
+	relabel []*relabel.Config,
+	jobName string,
+	config *scrapeconfig.GcplogTargetConfig,
+	pubsubClient *pubsub.Client,
+	ctx context.Context,
+	cancel func(),
+) (*GcplogTarget, error) {
+
+	pt := &GcplogTarget{
+		logger:        logger,
+		handler:       handler,
+		relabelConfig: relabel,
+		config:        config,
+		jobName:       jobName,
+		ctx:           ctx,
+		cancel:        cancel,
+		ps:            pubsubClient,
+		msgs:          make(chan *pubsub.Message),
+	}
+
+	return pt, nil
+}
+
+func (t *GcplogTarget) run() error {
+	send := t.handler.Chan()
+
+	sub := t.ps.SubscriptionInProject(t.config.Subscription, t.config.ProjectID)
+	go func() {
+		err := sub.Receive(t.ctx, func(ctx context.Context, m *pubsub.Message) {
+			t.msgs <- m
+		})
+		if err != nil {
+			// TODO(kavi): Add proper error propagation maybe?
+			level.Error(t.logger).Log("error", err)
+			gcplogErrors.WithLabelValues(t.config.ProjectID).Inc()
```

Comment:
something is off here. This can exit with an error and the main for {} will be waiting but no message will ever show up.

I think it should just at least let know somehow that this is not running, may be via the Ready & Details to show the error ? 

- Should we just stop everything ? 
- Should we retry ?

May be that's for later

Reply:
I think this error (coming from pubsub Receive) can mean two things.
- Some transient errors which can be retried
- Permanent error (may be some invalid configs or anything). 

In case of 2, we cannot do much other than need some human attention (via some alerts or anything). Killing or stopping everything may also won't work (it restarts and same issue remains). 

But I think these we will have clear picture once we start testing more on dev. Sure that can be done later.

Reply:
my point was this goroutine will die if an error is thrown.

Meaning it will never receive message. shouldn't the target this information ?

Reply:
It won't retry automatically to receive pubsub message, so it is in a state where it's stuck, may be in that case it should not be ready.

Reply:
Ah I see what you mean. 

So I think its better to defer ctx.Cance() in this goroutine?. So that main loop also exits from `run`. The more I think about it. we can't do much if `sub.Receive` fails so that would be good option. wdyt?

Reply:
Added it!